apps/opt.c

```diff
@@ -656,22 +659,54 @@ int opt_next(void)
             /* Just a string. */
             break;
         case '/':
-            if (app_isdir(arg) >= 0)
+            if (app_isdir(arg) > 0)
                 break;
             BIO_printf(bio_err, "%s: Not a directory: %s\n", prog, arg);
             return -1;
         case '<':
             /* Input file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) >= 0)
+            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) == 0)
                 break;
             BIO_printf(bio_err,
                        "%s: Cannot open input file %s, %s\n",
                        prog, arg, strerror(errno));
             return -1;
         case '>':
             /* Output file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, W_OK) >= 0 || errno == ENOENT)
+            l = strlen(arg);
+            if (arg[l - 1] == '/' || app_isdir(arg) > 0) {
```

Comment:
nope, checking the parent is enough, since if `/tmp/non-exist1/non-exist2` is not writable, for whatever reason, we don't need to check anymore, the whole place that one wants to write a file in is not valid. The logic is: If one wants to write a file in a location, he/she must have the write permission of the location(must be a directory in Unix-like system) and might need to also have the write permission for the intented-file if the file is already there.

The current logic's problem is, only checking `errno == ENOENT` is too *BIG* to cover the `file-not-exist-but-directory-is-exist-and-writable' case, it's also tolerant to other 'should-not-be-tolerant' cases, and this is what #3404 reported.

(sorry for typing some 'stick conjunctional' words, since I found this is a easy way to express in English : )